### PR TITLE
`Profile`: do not leak repository implementation in interface

### LINKED ISSUE #4890: Do not leak disk-objectstore specificities in the public interface of the `Profile`

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #4890 

The `Profile` interface was leaking implementation details of the
repository. For example, the method `get_repository_container` returned
the `Container` class of the `disk-objectstore` library, but this is an
implementation detail. If this ever were to change all client code would
break.

Instead, we add the `get_repository` method which returns an instance of
the generic `Repository` class that will be common to all repository
implementations. The backend implementation can still be obtained
through this object but this should only be done in exceptional cases.

The `Repository` class now has three additional properties and methods:

 * `def uuid(self) -> typing.Optional[str]:`
 * `def initialise(self, **kwargs) -> None:`
 * `def is_initialised(self) -> bool:`

These simply call through to the exact same method/property on the
backend instance. The `AbstractRepositoryBackend` also now has the exact
same attributes and they are implemented for the two currently existing
implementations `DiskObjectStoreRepositoryBackend` and the
`SandboxRepositoryBackend`.

### CODE REVIEW

**Comment by sphuber** [comment]
Haven't gone over all the comments of @CasperWA of the other PR yet, but will do later. I propose @mbercx already goes over the suggested interface changes here. Will have time to get back to this later in the afternoon.

**Review by chrisjsewell — APPROVED** [review]

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
well perhaps I should say, approved after it doesn't break all the tests lol

**Comment by CasperWA** [comment]
> Haven't gone over all the comments of @CasperWA of the other PR yet, but will do later (...)

They're all syntax and style, essentially. So no worries.

File: aiida/manage/configuration/profile.py
Comment (chrisjsewell) [inline aiida/manage/configuration/profile.py:132]: paths should be return as `pathlib.Path`
if not already the case, it would be good if `Container` accepted a `Path` for its initialisation

Comment (CasperWA) [reply aiida/manage/configuration/profile.py:132]: One of my comments as well 😅

Comment (mbercx) [reply aiida/manage/configuration/profile.py:132]: In my (very poor) defence, I only copy-pasted @giovannipizzi's code snippets from slack and tested if it worked. 😅

Comment (CasperWA) [reply aiida/manage/configuration/profile.py:132]: Those are not meant as the final version to go into the codebase I'm suspecting ;) So that's a "no" on your defence 😂

Comment (CasperWA) [reply aiida/manage/configuration/profile.py:132]: Also, @mbercx if you haven't already, you should install `pre-commit` in the repo by running `pre-commit install`. If it says you don't have `pre-commit` install with the `pre-commit` extra, i.e., `pip install -U -e .[pre-commit]` (or similar).

File: aiida/manage/configuration/profile.py
Comment (mbercx) [inline aiida/manage/configuration/profile.py:142]:
```suggestion
        """Return the repository configured for this profile.
        
        .. note:: will initialise the repository if not already done so.
        """
```

File: aiida/manage/manager.py
Comment (mbercx) [inline aiida/manage/manager.py:139]: Is this check still necessary in case `get_repository_uuid()` returns `None` for non-initialised repositories?

File: aiida/manage/configuration/profile.py
Comment (mbercx) [inline aiida/manage/configuration/profile.py:153]:
```suggestion
        .. note:: will return ``None`` if the repository has not yet been initialised.
```
Else @chrisjsewell will be angry. 😁

File: aiida/manage/manager.py
Comment (mbercx) [reply aiida/manage/manager.py:139]: If not, perhaps we should make this a hidden method, or even remove it entirely? I feel we may be overdoing it with the number of repository methods/properties. 😅

**Review by mbercx — CHANGES_REQUESTED** [review]
Thanks @sphuber. Gave the PR a first pass.

File: aiida/manage/configuration/profile.py
Comment (mbercx) [reply aiida/manage/configuration/profile.py:132]: I have the pre-commit hooks installed, and they all passed for my commit. 🤷‍♂️  I haven't run `pre-commit install` in a while though, maybe the hooks have changed since then? I didn't know there should be an empty line after an import in a method (I'm assuming you're referring to [this comment](https://github.com/aiidateam/aiida-core/pull/4889#discussion_r622917362)).

Comment (CasperWA) [reply aiida/manage/configuration/profile.py:132]: Indeed. I thought we were using black now. i guess not? But black always puts an empty new line after imports.

Comment (sphuber) [reply aiida/manage/configuration/profile.py:132]: the `Container` calls `os.path.realpath` on the path argument. Does that support `Path` objects already in Python 3.6? If so, then we are good.

File: aiida/manage/manager.py
Comment (sphuber) [reply aiida/manage/manager.py:139]: Maybe what would be better is that, now `Profile.get_repository` returns a `Repository` instance, we implement those methods there. That should be the correct way of doing things.

File: aiida/manage/configuration/profile.py
Comment (chrisjsewell) [reply aiida/manage/configuration/profile.py:132]: > Edit: Ah, no. We're indeed not yet using black. Still on yapf.... pff

Ah you are preaching to the choir my friend 😆: https://github.com/aiidateam/aiida-core/pull/4345#issuecomment-826762603

> you should install pre-commit in the repo by running pre-commit install

you could really do that with aiida-core, because pylint and mypy use system installs: use tox (its the way forward): `tox -e py38-pre-commit`

Comment (chrisjsewell) [reply aiida/manage/configuration/profile.py:132]: >  Does that support Path objects already in Python 3.6?

I think everything in `os.path` supports `Path` for >=3.6, but can check

Comment (chrisjsewell) [reply aiida/manage/configuration/profile.py:132]: yep:

```python
In [5]: realpath??
Signature: realpath(filename)
Source:   
def realpath(filename):
    """Return the canonical path of the specified filename, eliminating any
symbolic links encountered in the path."""
    filename = os.fspath(filename)
    path, ok = _joinrealpath(filename[:0], filename, {})
    return abspath(path)
File:      ~/opt/miniconda3/envs/python-3.6/lib/python3.6/posixpath.py
Type:      function

In [6]: realpath(Path("."))
Out[6]: '/Users/chrisjsewell/Documents/GitHub/aiida-core-origin'
```

File: aiida/manage/manager.py
Comment (mbercx) [reply aiida/manage/manager.py:139]: Indeed, having `Repository.is_initialised` and `Repository.initialise` sounds like a good idea. 👍

**Comment by sphuber** [comment]
Alright, this should be much better. If you guys are happy with the design, I will add the necessary tests for the new `Repository` and `RepositoryBackend` properties and methods.

**Comment by mbercx** [comment]
> Alright, this should be much better. If you guys are happy with the design, I will add the necessary tests for the new `Repository` and `RepositoryBackend` properties and methods.

Looking very good indeed! Much cleaner to have these methods on the `Repository` class. 👌  Will do a proper review and some testing after the wrap-up meeting.

**Comment by chrisjsewell** [comment]
> Looking very good indeed!

not that it means much to have @mbercx 's seal of a approval

**Comment by CasperWA** [comment]
> > Looking very good indeed!
> 
> not that it means much to have @mbercx 's seal of a approval

Indeed. I quickly jumped in to review when I saw I blanketed this PR ... tsk tsk tsk... 😂

**Comment by mbercx** [comment]
> > Looking very good indeed!
> 
> not that it means much to have @mbercx 's seal of a approval

Does that mean I don't have to review PR's anymore? 🥳

File: aiida/repository/backend/sandbox.py
Comment (CasperWA) [inline aiida/repository/backend/sandbox.py:23]: The UUID is not being returned here, might be nice? Also whether or not it's initialised?

File: aiida/repository/repository.py
Comment (CasperWA) [inline aiida/repository/repository.py:52]: See comment concerning the string representation of `SandboxRepositoryBackend`.

File: tests/cmdline/commands/test_setup.py
Comment (CasperWA) [inline tests/cmdline/commands/test_setup.py:83]: It reads a bit redundant now I think. But up to you.
```suggestion
        self.assertEqual(backend_manager.get_repository_uuid(), profile.get_repository().uuid)
```

File: tests/cmdline/commands/test_setup.py
Comment (CasperWA) [inline tests/cmdline/commands/test_setup.py:171]:
```suggestion
        self.assertEqual(backend_manager.get_repository_uuid(), profile.get_repository().uuid)
```

File: tests/cmdline/commands/test_setup.py
Comment (CasperWA) [reply tests/cmdline/commands/test_setup.py:83]: If it's used later as a proper variable fine to keep (can't see that it is). The same goes for the suggested change below.

**Comment by CasperWA** [comment]
> > > Looking very good indeed!
> > 
> > not that it means much to have @mbercx 's seal of a approval
> 
> Does that mean I don't have to review PR's anymore? 🥳

Nah. Just means more work for the rest of us 😂

File: aiida/cmdline/commands/cmd_status.py
Comment (sphuber) [reply aiida/cmdline/commands/cmd_status.py:93]: It does now :) I implemented a `__str__` on the repository implementations and the `DiskObjectStore` one includes its UUID as well as absolute filepath

File: aiida/repository/backend/sandbox.py
Comment (sphuber) [reply aiida/repository/backend/sandbox.py:23]: Sandbox doesn't have a UUID and always returns `None`. It doesn't make sense for a sandbox tbh

File: aiida/repository/repository.py
Comment (sphuber) [reply aiida/repository/repository.py:52]: Since this calls `str(self.backend)` this will include whatever the `__str__` of the backend class implements. So for the DOS one, it will include the UUID.

File: aiida/repository/backend/sandbox.py
Comment (CasperWA) [reply aiida/repository/backend/sandbox.py:23]: Ah right. Yeah. There were so many layers here I kind of lost track of which string is represented where 😅

**Comment by sphuber** [comment]
Ok, the docs are failing with the following:
```
/home/circleci/.local/lib/python3.7/site-packages/aiida/manage/configuration/profile.py:docstring of aiida.manage.configuration.Profile.get_repository:: WARNING: py:class reference target not found: Repository
/home/circleci/.local/lib/python3.7/site-packages/aiida/manage/configuration/profile.py:docstring of aiida.manage.configuration.profile.Profile.get_repository:: WARNING: py:class reference target not found: Repository
```
This is most likely becuase of the return type, which I put as a string. The reason being that I cannot import `Repository` at the top of the module, because this leads to a circular import. For the typing, I put a guard in `if typing.TYPE_CHECKING` which makes sure `mypy` can properly import it, but for the docs this doesn't work. Not sure how to get around this. Any ideas? Other than maybe trying to break the circular import in the `aiida.repository.Repository` path, but that might not be trivial.

**Comment by chrisjsewell** [comment]
> This is most likely becuase of the return type, which I put as a string. 

yeh its an annoying sphinx thing (https://github.com/sphinx-doc/sphinx/issues/8498), just add it to the `docs/source/nitpick-exceptions`

**Review by chrisjsewell — APPROVED** [review]

**Comment by sphuber** [comment]
@mbercx would you still like a gander before we merge?

**Comment by mbercx** [comment]
> @mbercx would you still like a gander before we merge?

LGMT! 🚀

**Pull request merged**
